lib/rouge/lexers/mmt.rb

```diff
@@ -0,0 +1,254 @@
+# -*- coding: utf-8 -*- #
+# frozen_string_literal: true
+
+module Rouge
+  module Lexers
+    class MMT < RegexLexer
+      title 'mmt'
+      desc 'MMT Surface Syntax (uniformal.github.io)'
+      tag 'mmt'
+      aliases 'mmtx'
+      mimetypes 'application/x-mmt'
+      filenames '*.mmt','*.mmtx'
+
+      state :root do
+        rule %r/\s+/im, Text::Whitespace
+        rule %r/\/T .*?\u275a/im, Comment::Multiline
+        rule %r/\/\/.*?\u275a/im, Comment::Multiline
+
+        rule %r/(document)([ \t]+)(\S+)/im do
+          groups Keyword::Declaration, Text::Whitespace, Name::Namespace
+        end
+
+        rule %r/(import)(\s+)(\S+)/im do
+          groups Keyword::Namespace, Text::Whitespace, Name::Namespace
+        end
+
+        rule %r/meta\b/im, Keyword::Declaration
+        rule %r/namespace\b/im, Keyword::Namespace
+        rule %r/(fixmeta|ref|rule)\b/im, Comment::Preproc
+        rule %r/(total|implicit)\b/im, Keyword
+
+        rule %r/theory\b/im, Keyword::Declaration, :theoryHeader
+        rule %r/view\b/im, Keyword::Declaration, :viewHeader
+        rule %r/diagram\b/im, Keyword::Declaration, :diagramHeader
+
+        rule %r/./, Text
+      end
+
+      state :theoryHeader do
+        rule %r/\s+/im, Text::Whitespace
+
+        rule %r/[^\s:=>]+/im, Name::Class
+
+        rule %r/(:)(\s*)([^\s=]+)/im do
+          groups Punctuation, Text::Whitespace, Text
+        end
+
+        rule %r/(>)([^=]+)/im do
+          groups Punctuation, Name::Variable
+        end
+
+        rule(//) { goto :moduleDefines }
```

Comment:
I'm not sure if Pygments supports this feature but Rouge allows you to essentially 'move' from one state to another using `goto` (really it pops the current state and then pushes the target state onto the stack). This is usually preferable because it means that subsequent states don't need to call `pop!` multiple times (which can lead to errors if you pop more states than are on the stack).

Oh, and as you may have guessed `rule (//)` will match anything without advancing the current character at all. This is a way to ensure that the lexer will move from state to another.

Reply:
Actually, I don't know whether Pygments supports moving without pushing the current state. 

> This is usually preferable because it means that subsequent states don't need to call pop! multiple times

But then, they can never return to the original state, can they? That is without knowing the original state being hardcoded in the general design.

In the MMT language theories (which are modules) can be highly nested ([see in sample here](https://github.com/rouge-ruby/rouge/blob/eedd87a005f8283866651b6317be739bcd812457/spec/visual/samples/mmt)). Hence, I designed the lexer to make heavy used of pushing and popping. I wouldn't know how to implement it otherwise.

Reply:
> But then, they can never return to the original state, can they? That is without knowing the original state being hardcoded in the general design.

That's correct, but the situations where I used calls to `goto` were to states that did not return to the immediately previous state. Rather they returned to earlier states (which they accomplished by calling `pop!` with a number greater than 1). Or at least that was my intent. Did I screw that up somewhere?